clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionSimulationCoordinator.java

```diff
@@ -0,0 +1,316 @@
+            );
+        } else if (request.data.producerEpoch() != currentEpoch) {
+            return new AddOffsetsToTxnResponse(
+                new AddOffsetsToTxnResponseData()
+                    .setErrorCode(Errors.INVALID_PRODUCER_EPOCH.code())
+                    .setThrottleTimeMs(throttleTimeMs)
+            );
+        } else {
+            offsetsAddedToTxn = true;
+            return new AddOffsetsToTxnResponse(
+                new AddOffsetsToTxnResponseData()
+                    .setErrorCode(Errors.NONE.code())
+                    .setThrottleTimeMs(throttleTimeMs)
+            );
+        }
+    }
+
+    private AbstractResponse handleTxnCommit(TxnOffsetCommitRequest request,
+                                             final boolean faultInject) {
+        Map<TopicPartition, Errors> errors = new HashMap<>();
+        request.data.topics().forEach(topic -> topic.partitions().forEach(partition -> {
+            TopicPartition key = new TopicPartition(topic.name(), partition.partitionIndex());
+            if (faultInject) {
+                errors.put(key, Errors.COORDINATOR_LOAD_IN_PROGRESS);
+            } else if (request.data.producerId() != currentProducerId) {
+                errors.put(key, Errors.UNKNOWN_PRODUCER_ID);
+            } else if (request.data.producerEpoch() != currentEpoch) {
+                errors.put(key, Errors.INVALID_PRODUCER_EPOCH);
+            } else if (offsetsAddedToTxn) {
+                pendingOffsets.put(key, partition.committedOffset());
+                errors.put(key, Errors.NONE);
+            } else {
+                errors.put(key, Errors.UNKNOWN_TOPIC_OR_PARTITION);
+            }
+        }));
+
+        return new TxnOffsetCommitResponse(
+            throttleTimeMs,
+            errors
+        );
+    }
+
+    private AbstractResponse handleProduce(ProduceRequest request,
+                                           final boolean faultInject) {
+        Map<TopicPartition, PartitionResponse> errors = new HashMap<>();
+        Map<TopicPartition, MemoryRecords> partitionRecords = request.partitionRecordsOrFail();
+
+        partitionRecords.forEach((topicPartition, records) -> {
+            if (faultInject) {
+                // Trigger KIP-360 path.
+                errors.put(topicPartition, new PartitionResponse(Errors.UNKNOWN_PRODUCER_ID));
+            } else {
+                List<Record> sentRecords = pendingPartitionData.getOrDefault(topicPartition, new ArrayList<>());
+                for (Record partitionRecord  : records.records()) {
+                    sentRecords.add(partitionRecord);
+                }
+
+                pendingPartitionData.put(topicPartition, sentRecords);
+                errors.put(topicPartition, new PartitionResponse(Errors.NONE));
+            }
+        });
+
+        return new ProduceResponse(errors, throttleTimeMs);
+    }
+
+    private EndTxnResponse handleEndTxn(EndTxnRequest request, final boolean faultInject) {
```

Comment:
Another class of failure that we can simulate is when a request reaches the broker and gets handled, but the connection is lost before the response is sent.

Reply:
Thought we do have that coverage on L126: 
```
        networkClient.respond(response, disconnect);
```